File: web/src/lib/labeldraft.ts
import { useQuery } from "@tanstack/solid-query";
import { api } from "../api/client";
import { describeError } from "./format";


// The draft identity render (#699, #702): what the console asks so a create
// form can show the name AND the label the platform is about to write, in
// locked fields, before the row exists.
//
// The label cannot be rendered in the browser and that is a decision rather
// than a gap (ADR-0098, ADR-0104): a rule is a Go text/template over a closed
// data map behind an AST allowlist, and a second implementation of it in
// TypeScript is exactly the defect slice 3 of this epic swept 42 hand-rolled
// copies of. The NAME used to be the exception, resolved in the browser by
// walking the type chain for a stem and writing a token where the ordinal went;
// #702 removed that, because the server had already resolved all three to
// render the label and returning them costs nothing. So the one engine answers
// both halves, over the wire, and this module is the call rather than the
// calculation.
//
// What makes that affordable is the same thing that makes it honest: the route
// allocates nothing. It reads the lowest free ordinal rather than minting one,
// takes no advisory lock and opens no write transaction, so asking on every
// picker change costs a read, not a place in the queue every create in the
// estate is waiting in.

// DraftLabel is the answer: the name, the ordinal that name was minted from,
// the label, and the rule that produced it.
//
// The ordinal is what a form posts back as expected_ordinal, and it is absent
// exactly when the caller supplied the name, because an operator-named row
// carries no ordinal at all. So "the field is locked" and "there is a
// precondition to post" are one fact rather than two.
//
// An empty label is a real state rather than an error: it means the platform
// stores nothing and the surface reads the name instead. It is reached by
// clearing the rule at every tier, rather than being where a shipped estate
// starts (each of the three kinds ships a rule as of #657). The rule field is
// what tells it apart from a rule that ran and had nothing to say.
export interface DraftLabel {
  name: string;
  ordinal?: number;
  label: string;
  rule: string;
}

// The three draft bodies. Each mirrors its own create body field for field, and
// the NAME field carries the same meaning in both: omitted is the operator
// handing the platform the pen. That is what keeps a locked field and what gets
// posted from disagreeing, since the form asks both routes the same question in
// the same shape and posts the same object to each.
//
// parent is on all three (#702) because it is half of the placement bucket the
// previewed ordinal is read from: a draft that left it out would preview the
// wrong bucket and the create beside it would be refused by the very
// precondition meant to protect it.
export interface ComponentDraft {
  product: string;
  name?: string;
  parent?: string;
  location?: string;
  system?: string;
}

export interface SystemDraft {
  system_type_id?: string;
  standard_id?: string;
  name?: string;
  parent?: string;
  location?: string;
}

export interface LocationDraft {
  location_type: string;
  name?: string;
  parent?: string;
}

export type EstateDraft =
  | { kind: "component"; body: ComponentDraft }
  | { kind: "system"; body: SystemDraft }
  | { kind: "location"; body: LocationDraft };

export async function renderComponentLabel(body: ComponentDraft): Promise<DraftLabel> {
  const { data, error } = await api.POST("/components:renderLabel", { body });
  if (error) throw error;
  return data as DraftLabel;
}

export async function renderSystemLabel(body: SystemDraft): Promise<DraftLabel> {
  const { data, error } = await api.POST("/systems:renderLabel", { body });
  if (error) throw error;
  return data as DraftLabel;
}

export async function renderLocationLabel(body: LocationDraft): Promise<DraftLabel> {
  const { data, error } = await api.POST("/locations:renderLabel", { body });
  if (error) throw error;
  return data as DraftLabel;
}

export function renderDraftLabel(d: EstateDraft): Promise<DraftLabel> {
  switch (d.kind) {
    case "component":
      return renderComponentLabel(d.body);
    case "system":
      return renderSystemLabel(d.body);
    default:
      return renderLocationLabel(d.body);
  }
}

// LABEL_DRAFT_KEY namespaces the draft in the query cache. The body itself is
// the rest of the key, so changing any picker asks a new question and an
// unchanged form asks none: the render is a pure function of the body, which is
// what makes it cacheable at all and is the same property that lets it skip the
// lock.
export const LABEL_DRAFT_KEY = "label-draft";

// useLabelDraft renders the label for whatever the form currently holds, or
// nothing at all when the form is not ready to be asked.
//
// draft returning null is "do not ask": an unchosen product, an unchosen
// location type, or a name field the operator has unlocked and left empty. The
// query is disabled rather than fired with a half-filled body, because a 422
// from a form mid-edit is noise an operator would read as a failure.
//
// retry is off. The refusals this route gives are 422s about the body (no stem,
// no name rule, a placement out of scope), and none of them gets better by
// being asked again; a retry would only delay the message that names the
// missing fact.
export function useLabelDraft(draft: () => EstateDraft | null) {
  return useQuery(() => {
    const d = draft();
    return {
      queryKey: [LABEL_DRAFT_KEY, d?.kind, d?.body] as const,
      queryFn: () => renderDraftLabel(d!),
      enabled: d !== null,
      retry: false,
      staleTime: 30_000,
    };
  });
}

// The two refusals a create form has to ACT on, read out of the RFC 9457
// errors array rather than out of the message (#702).
//
// Every other refusal these routes give means "show it and stop". These two
// have a specific recovery, and matching on the sentence to find them would tie
// the console to server copy that is free to change. The server stamps each
// with the request field the recovery acts on, which is a machine-readable
// channel the ErrorModel already publishes and the generated client already
// types.
interface ProblemDetail {
  location?: string;
  value?: unknown;
}

function details(e: unknown): ProblemDetail[] {
  const errs = (e as { errors?: ProblemDetail[] | null } | null | undefined)?.errors;
  return Array.isArray(errs) ? errs : [];
}

// nameRefused is "the platform will not name this row": a component_type chain
// with no stem, an unclassified system, a system_type chain with no stem, or a
// location_type with no name rule. All four are permanent answers rather than
// loading states, and all four mean the same thing to the form, which is that
// the operator has to type a name.
export function nameRefused(e: unknown): boolean {
  return details(e).some((d) => d.location === "body.name");
}

// ordinalTaken is "another create took the number this form was holding",
// returning the ordinal that moved so a surface can say what happened without
// waiting for the re-read. Null when the refusal is anything else, including
// the OTHER 409 a create can give (a name collision), whose recovery is the
// operator's rather than the form's.
export function ordinalTaken(e: unknown): number | null {
  const d = details(e).find((x) => x.location === "body.expected_ordinal");
  return typeof d?.value === "number" ? d.value : null;
}

// recoverFromTakenOrdinal is what a create form does with a refused submit, and
// it is one function rather than three because the recovery is the same on every
// tier: re-read the draft so the locked field shows the name that is free now,
// and hand back the sentence to put in front of the operator.
//
// Everything else falls through unchanged, which is the point of asking
// ordinalTaken rather than the status: a name collision is a 409 too, and
// re-reading the draft would tell that operator nothing.
//
// The extra sentence is the console's, not the server's: the server says what
// happened to the request, and only the form knows that it has just refreshed a
// field further down the page.
export async function recoverFromTakenOrdinal(
  e: unknown,
  refetch: () => Promise<unknown>,
): Promise<string> {
  if (ordinalTaken(e) === null) return describeError(e);
  await refetch();
  return `${describeError(e)} The name below now shows what is free; create again to take it.`;
}

File: internal/storage/label_draft.go
package storage

import (
	"context"
	"errors"
	"fmt"

	"github.com/hyperscaleav/omniglass/internal/label"
	"github.com/hyperscaleav/omniglass/internal/scope"
	"github.com/jackc/pgx/v5"
)

// The draft identity render (#699, #702): the name and the label a row WOULD
// carry, answered before the row exists, for a create form that wants to show
// the operator what the platform is about to name and label rather than only
// promising that it will.
//
// # Why this is not the preview ADR-0104 refused
//
// That refusal is about MINTING. A draft preview that allocated an ordinal took
// the same pg_advisory_xact_lock on the placement bucket that real creates
// take, so a form previewing per picker change would serialise the fleet's
// creates behind a UI affordance.
//
// A render allocates nothing. It resolves the rule through the same tiers, over
// the same closed data map, with the same one engine, and it READS the lowest
// free ordinal in the placement bucket ([previewName]) rather than allocating
// one. No lock, no write transaction, no allocation, and no second
// implementation of anything: every function it calls is the one the create's
// own stamp calls.
//
// The other half of that refusal was that the answer is provisional, and it is:
// another create can take the number in between. That is answered by the
// PRECONDITION rather than by hiding the number (#702). The form posts the
// ordinal it was shown, the create compares it with what it allocated
// (confirmOrdinal), and a create that would land a different name is refused
// with the number that moved. An operator is either given the name they were
// shown or told it changed, and never quietly handed a third thing.
//
// # What it reuses, and why that is the point
//
// componentLabelChain / systemLabelChainWith / locationLabelChainWith resolve
// the tiers; componentLabelData / systemLabelData / locationLabelData build the
// closed map (which IS the sandbox: a key absent there is a key a rule cannot
// reach, and this file adds none); renderLabel executes and degrades;
// previewName reads the ordinal from the same siblings the allocator tests, with
// the same mint. What the draft supplies differently is only the two facts a row
// that does not exist cannot have of its own, and it now supplies BOTH from the
// same primitives the create uses rather than leaving either as a token.
//
// # Placement, and the scopes that guard it
//
// The draft takes the same scopes its create takes, in the same order, because
// it resolves the same references: a component's parent within the caller's
// component:create scope (the set the create resolves it in), its location
// within location:read (the set that decides whether the rendered string may
// carry that label), and its system within system:read AND system:update, which
// is the set the create's bind resolves in because that bind writes the system's
// membership (#707, #713). An out-of-scope reference is the same non-disclosing
// not-found a direct read would give. Rendering it anyway would make this route
// a disclosure channel, and resolving it in a WIDER set than the create's would
// make it a preview of a create the platform refuses.
//
// The parent is new here (#702) and it is not about the label at all: a
// location's map carries no placement (ADR-0098's exclusion survives on that
// tier), but the ordinal is read from a BUCKET, and the bucket is exactly the
// parent, else the location, else the unplaced one. So a location's draft now
// takes a scope where it took none, and what it guards is the sibling read
// rather than the render.

// DraftLabel is what a create form is told about the identity it is about to
// produce: the name, the ordinal that name was minted from, the rendered label,
// and the rule that rendered it.
//
// The NAME is here (#702) rather than resolved a second time in the browser
// because the label is rendered FROM it: the server has already resolved the
// stem, the mint and the ordinal to answer at all, so returning them costs
// nothing and removes the console's need to know what a mint looks like. It is
// also the only way the two halves of a create form can be one answer rather
// than two that can disagree.
//
// The rule travels with the answer for two reasons. It is the only way to tell
// "no rule applies at any tier" (both label fields empty, which an operator
// reaches by clearing the rule at every tier) from "a rule applies and had
// nothing to say about this row" (a rule, an empty label). And a surface that
// shows where a value came from teaches the mechanism it operates, which a bare
// string cannot.
type DraftLabel struct {
	// Name is the name the create would stamp: the one the caller supplied, or
	// the one the platform would mint at Ordinal.
	Name string
	// Ordinal is the number Name was minted from, and 0 when the caller
	// supplied the name, because that row carries no ordinal at all (#681). It
	// is what a form posts back as its precondition, and 0 is deliberately not
	// a legal value to post: absent and zero are the same state, which is "the
	// platform allocated nothing here".
	Ordinal int
	// Label is the rendered label, or "" when no rule resolves or the rule
	// renders nothing. Empty is not a failure: the read ladder's third rung is
	// the entity's own name, so an empty label means the name shows instead.
	Label string
	// Rule is the template that produced Label, resolved through the same tier
	// precedence a create uses, or "" when no tier had an opinion.
	Rule string
}

// ComponentLabelDraft is a component that does not exist yet. Name empty is the
// operator handing the pen to the platform, exactly as it is on ComponentSpec,
// so the draft and the create it describes are asked the same question in the
// same shape and cannot be given different answers by a caller that fills one
// field and not the other.
type ComponentLabelDraft struct {
	ProductName string
	Name        string
	// ParentName, LocationName and SystemName mirror the create's own placement
	// fields. The first two are also the NAME's placement bucket (a parent wins
	// over a location, and neither is the unplaced bucket), which is what the
	// previewed ordinal is read from: a form that showed the ordinal of a
	// bucket the create does not write into would be worse than showing none.
	ParentName   string
	LocationName string
	SystemName   string
}

// SystemLabelDraft is ComponentLabelDraft on the system tier.
type SystemLabelDraft struct {
	SystemTypeRef string
	StandardRef   string
	Name          string
	ParentName    string
	LocationName  string
}

// LocationLabelDraft is ComponentLabelDraft on the location tier. Its LABEL
// carries no placement: a location's data map is its name and its type, and
// deliberately nothing about where it sits (ADR-0098's exclusion survives on
// this tier). Its NAME does: a location has two placement buckets, under a
// parent or at the root, so ParentName is what the previewed ordinal is read
// from.
type LocationLabelDraft struct {
	LocationTypeRef string
	Name            string
	ParentName      string
}

// RenderComponentDraftLabel renders the name and the label a component create
// would stamp, without creating anything.
//
// The scopes are the create's own, in the create's own order and now in the
// create's own number (#713): create resolves the PARENT (the set
// CreateComponent resolves it in, and the bucket the previewed ordinal is read
// from), locationRead the location whose label the map reads, and systemRead
// with systemUpdate the system, through the same read-then-action split the
// create's bind uses. The permission to ASK is the component's own :create and
// lives at the route (see internal/api/label_draft.go); these four decide what
// the answer may contain.
func (p *PG) RenderComponentDraftLabel(ctx context.Context, d ComponentLabelDraft, create, locationRead, systemRead, systemUpdate scope.Set) (DraftLabel, error) {
	productID, err := registryID(ctx, p.pool, "product", d.ProductName, ErrProductNotFound)
	if err != nil {
		return DraftLabel{}, err
	}
	in, err := componentLabelChain(ctx, p.pool, productID)
	if err != nil {
		return DraftLabel{}, err
	}
	// Placement first, because the NAME is read from it now: the ordinal is the
	// lowest free one among the siblings in this exact bucket, so a draft that
	// resolved the placement after minting would be answering about a different
	// bucket than the create writes into.
	parentID, err := draftParentID(ctx, p.pool, componentConfig, d.ParentName, "component", create, ErrComponentNotFound, ErrParentComponentNotFound)
	if err != nil {
		return DraftLabel{}, err
	}
	var (
		pl         componentPlacement
		locationID *string
	)
	if d.LocationName != "" {
		loc, err := p.GetLocation(ctx, d.LocationName, locationRead)
		if err != nil {
			return DraftLabel{}, err
		}
		pl.locationLabel = locationReadLabel(loc)
		locationID = &loc.ID
	}
	if d.SystemName != "" {
		// resolvePlacementWriteRef, the create's own resolver, not a read (#713).
		// A component create INSERTS this system's membership, so the create
		// requires system:update over it and resolves it there; a draft that
		// resolved the same reference in system:read alone would preview a create
		// the platform then refuses, and the preview would carry that system's
		// type label back out (the disclosure the read scope exists to close, one
		// tier further along). The read set still decides what the refusal may
		// SAY, exactly as it does on the create.
		sys, err := resolvePlacementWriteRef(ctx, p.pool, systemConfig, d.SystemName, systemRead, systemUpdate)
		if errors.Is(err, ErrSystemForbidden) {
			return DraftLabel{}, ErrSystemBindForbidden
		}
		if err != nil {
			return DraftLabel{}, err
		}
		if pl.systemTypeLabel, err = systemTypeLabelOf(ctx, p.pool, sys.SystemTypeID); err != nil {
			return DraftLabel{}, err
		}
	}
	c := Component{Name: d.Name}
	if d.Name == "" {
		// The same refusal generateNameForProduct raises, from the same
		// resolved stem: a form must never lock a field over a value the
		// create then declines to produce. componentLabelChain has already
		// walked the chain for the map's Stem key, so this costs nothing.
		if in.stem == "" {
			return DraftLabel{}, fmt.Errorf("%w: product %q", ErrComponentTypeNoStem, d.ProductName)
		}
		name, ordinal, err := previewName(ctx, p.pool, componentMint(in.stem), componentNameScope(parentID, locationID))
		if err != nil {
			return DraftLabel{}, err
		}
		c.Name, c.Ordinal = name, &ordinal
	}
	eng, err := p.labelEngine(ctx, p.pool)
	if err != nil {
		return DraftLabel{}, err
	}
	return draftedLabel(eng, in.rule, componentLabelData(&c, in, pl), c.Name, c.Ordinal), nil
}

// RenderSystemDraftLabel renders the name and the label a system create would
// stamp. create resolves the parent (the name's bucket), locationRead the one
// placement fact a system's map carries.
func (p *PG) RenderSystemDraftLabel(ctx context.Context, d SystemLabelDraft, create, locationRead scope.Set) (DraftLabel, error) {
	var systemTypeID, standardID *string
	if d.SystemTypeRef != "" {
		id, err := registryID(ctx, p.pool, "system_type", d.SystemTypeRef, ErrUnknownSystemType)
		if err != nil {
			return DraftLabel{}, err
		}
		systemTypeID = &id
	}
	if d.StandardRef != "" {
		id, err := registryID(ctx, p.pool, "standard", d.StandardRef, ErrUnknownStandard)
		if err != nil {
			return DraftLabel{}, err
		}
		standardID = &id
	}
	global, err := globalLabelRule(ctx, p.pool, "system")
	if err != nil {
		return DraftLabel{}, err
	}
	in, err := systemLabelChainWith(ctx, p.pool, standardID, systemTypeID, global)
	if err != nil {
		return DraftLabel{}, err
	}
	parentID, err := draftParentID(ctx, p.pool, systemConfig, d.ParentName, "system", create, ErrSystemNotFound, ErrParentSystemNotFound)
	if err != nil {
		return DraftLabel{}, err
	}
	var (
		pl         systemPlacement
		locationID *string
	)
	if d.LocationName != "" {
		loc, err := p.GetLocation(ctx, d.LocationName, locationRead)
		if err != nil {
			return DraftLabel{}, err
		}
		pl.locationLabel = locationReadLabel(loc)
		locationID = &loc.ID
	}
	s := System{Name: d.Name}
	if d.Name == "" {
		// generateNameForSystemType's two refusals, in its order: an
		// unclassified system has no registry row to take a stem from, and a
		// classified one whose chain sets none anywhere is a registry defect.
		// The console reads them as one state and says two different things,
		// because the operator's next move differs.
		if systemTypeID == nil {
			return DraftLabel{}, ErrSystemTypeRequiredForName
		}
		if in.stem == "" {
			return DraftLabel{}, fmt.Errorf("%w: system_type %s", ErrSystemTypeNoStem, *systemTypeID)
		}
		name, ordinal, err := previewName(ctx, p.pool, systemMint(in.stem), systemNameScope(parentID, locationID))
		if err != nil {
			return DraftLabel{}, err
		}
		s.Name, s.Ordinal = name, &ordinal
	}
	eng, err := p.labelEngine(ctx, p.pool)
	if err != nil {
		return DraftLabel{}, err
	}
	return draftedLabel(eng, in.rule, systemLabelData(&s, in, pl), s.Name, s.Ordinal), nil
}

// RenderLocationDraftLabel renders the name and the label a location create
// would stamp. Its one scope is the create's own, and it guards the PARENT
// rather than the render: a location's data map reads nothing off another
// fleet row (its keys are the location's own name and its type's display
// name), but the previewed ordinal is read from the parent's bucket, which is
// the same reference CreateLocation resolves in this same set.
func (p *PG) RenderLocationDraftLabel(ctx context.Context, d LocationLabelDraft, create scope.Set) (DraftLabel, error) {
	// Resolved first and unconditionally, so an unknown location_type is the
	// same ErrUnknownType a create gives rather than a label rendered with an
	// empty type name. locationLabelChainWith below tolerates a missing row on
	// purpose (a recompute must not stop on one), which is exactly the
	// tolerance a form must not inherit. It also has to be the uuid rather than
	// the operator-facing id, because that resolver keys on the column a stored
	// Location carries.
	locationTypeID, err := registryID(ctx, p.pool, "location_type", d.LocationTypeRef, ErrUnknownType)
	if err != nil {
		return DraftLabel{}, err
	}
	rule, err := locationNameRule(ctx, p.pool, locationTypeID)
	if err != nil {
		return DraftLabel{}, err
	}
	global, err := globalLabelRule(ctx, p.pool, "location")
	if err != nil {
		return DraftLabel{}, err
	}
	in, err := locationLabelChainWith(ctx, p.pool, locationTypeID, global)
	if err != nil {
		return DraftLabel{}, err
	}
	parentID, err := draftParentID(ctx, p.pool, locationConfig, d.ParentName, "location", create, ErrLocationNotFound, ErrParentNotFound)
	if err != nil {
		return DraftLabel{}, err
	}
	l := Location{Name: d.Name}
	if d.Name == "" {
		if rule == nil {
			return DraftLabel{}, fmt.Errorf("%w: location_type %s", ErrLocationTypeNoNameRule, locationTypeID)
		}
		name, ordinal, err := previewName(ctx, p.pool, rule.mint(), locationNameScope(parentID))
		if err != nil {
			return DraftLabel{}, err
		}
		l.Name, l.Ordinal = name, &ordinal
	}
	eng, err := p.labelEngine(ctx, p.pool)
	if err != nil {
		return DraftLabel{}, err
	}
	// A location's map deliberately carries no Ordinal key (labels.go), so the
	// previewed number reaches this tier's label only through the NAME the rule
	// reads, which is exactly how a stored location's own label is rendered.
	return draftedLabel(eng, in.rule, locationLabelData(&l, in), l.Name, l.Ordinal), nil
}

// draftParentID resolves the parent reference a drafted name's bucket is read
// from, within the caller's CREATE scope on that tier, and answers the
// parent-specific sentinel a create answers rather than the generic not-found.
//
// The create scope, not a read scope, and the difference matters: the reference
// is resolved here for the same reason CreateComponent resolves it, to decide
// which bucket the row lands in, so a draft resolving it in a wider set would
// preview an ordinal for a create the caller cannot make. The ordinal it
// discloses is one the caller could allocate for themselves by creating, which
// is what makes previewing it disclose nothing new.
//
// An empty ref is the PARENTLESS bucket, and it takes the create's own all-scope
// gate rather than resolving to nil unguarded (#702 review). Every one of the
// three creates refuses a parentless row to a caller whose create scope is not
// all (CreateLocation, CreateComponent, CreateSystem each answer cfg.forbidden),
// so a draft that answered it previewed a bucket the caller could never write
// into.
//
// That was a disclosure and not only an inconsistency. previewName reads the
// bucket's sibling NAMES and returns the lowest free ordinal, so the answer
// tells the caller which names in that bucket are taken. On the root bucket the
// probe is choosable: an operator holding location_type:create (seeded admin)
// forks a location type, writes any stem into its name rule, and asks the draft
// which ordinal is free for it. The function's own argument, that the caller
// could allocate the disclosed ordinal themselves by creating, is exactly what
// does not hold here.
//
// The gate is unconditional rather than only when a name is being drafted, for
// the same reason the create's is: what a caller cannot create, a form has no
// business rehearsing. The permission to ASK is still the route's gate; this is
// the scope, which is the other layer and enforced separately.
func draftParentID[T any](ctx context.Context, q querier, cfg scopedConfig[T], ref, resource string, create scope.Set, notFound, parentNotFound error) (*string, error) {
	if ref == "" {
		if !create.All {
			return nil, cfg.forbidden
		}
		return nil, nil
	}
	parent, err := resolveScopedRef(ctx, q, cfg, ref, resource, create)
	if errors.Is(err, notFound) {
		return nil, parentNotFound
	}
	if err != nil {
		return nil, err
	}
	id := cfg.idOf(parent)
	return &id, nil
}

// draftedLabel executes a resolved rule over a drafted map and returns it with
// the identity it was rendered for.
//
// Nothing is poked into the map any more (#702). It used to write the ordinal
// TOKEN over the Ordinal key, because the number was unknowable; the number is
// now read from the bucket before the map is built, so the drafted row carries
// it in the same field a stored row does and the three data builders stay what
// makes this safe: a key not added there is a key no rule can reach, and a
// draft adds none.
func draftedLabel(eng *label.Engine, rule string, data label.Data, name string, ordinal *int) DraftLabel {
	d := DraftLabel{Name: name, Label: renderLabel(eng, rule, data), Rule: rule}
	if ordinal != nil {
		d.Ordinal = *ordinal
	}
	return d
}

// registryID resolves a catalog reference (a name or a uuid) to the row's id,
// answering with the caller's own not-found sentinel so the refusal reads the
// way the create's does. Catalogs are not scoped trees, so there is no scope to
// inject here; the scoped refs a draft touches go through GetLocation and
// GetSystem instead.
//
// table is a compile-time constant at every call site, never operator input.
func registryID(ctx context.Context, q querier, table, ref string, notFound error) (string, error) {
	var id string
	err := q.QueryRow(ctx, `select id from `+table+` where `+registryRefCol(ref)+` = $1`, ref).Scan(&id)
	if errors.Is(err, pgx.ErrNoRows) {
		return "", notFound
	}
	if err != nil {
		return "", fmt.Errorf("storage: resolve %s %q: %w", table, ref, err)
	}
	return id, nil
}

// systemTypeLabelOf reads what SystemTypeLabel means for one system: its type's
// label, absent for an unclassified one. It is the single-row twin of
// the join componentPlacements does for a page, and the acceptance test (a
// drafted label compared with the one the create then stores, with a rule
// reading both placement facts) is what holds the two to the same answer.
func systemTypeLabelOf(ctx context.Context, q querier, systemTypeID *string) (string, error) {
	if systemTypeID == nil {
		return "", nil
	}
	var name string
	err := q.QueryRow(ctx, `select coalesce(label, '') from system_type where id = $1`, *systemTypeID).Scan(&name)
	if errors.Is(err, pgx.ErrNoRows) {
		return "", nil
	}
	if err != nil {
		return "", fmt.Errorf("storage: resolve system_type label %q: %w", *systemTypeID, err)
	}
	return name, nil
}
